action/staking/stake.go

```diff
@@ -0,0 +1,171 @@
+package staking
+
+import (
+	"encoding/json"
+
+	"github.com/Oneledger/protocol/action"
+	"github.com/Oneledger/protocol/data/keys"
+	"github.com/Oneledger/protocol/identity"
+	"github.com/pkg/errors"
+	"github.com/tendermint/tendermint/libs/kv"
+)
+
+var _ action.Msg = &Stake{}
+
+type Stake struct {
+	ValidatorAddress     keys.Address
+	StakeAddress         keys.Address
+	ValidatorPubKey      keys.PublicKey
+	ValidatorECDSAPubKey keys.PublicKey
+	NodeName             string
+	Stake                action.Amount
+}
+
+func (st Stake) Marshal() ([]byte, error) {
+	return json.Marshal(st)
+}
+
+func (st *Stake) Unmarshal(data []byte) error {
+	return json.Unmarshal(data, st)
+}
+
+func (st Stake) Signers() []action.Address {
+	return []action.Address{st.StakeAddress.Bytes(), st.ValidatorAddress.Bytes()}
+}
+
+func (st Stake) Type() action.Type {
+	return action.STAKE
+}
+
+func (st Stake) Tags() kv.Pairs {
+	tags := make([]kv.Pair, 0)
+
+	tag := kv.Pair{
+		Key:   []byte("tx.type"),
+		Value: []byte(st.Type().String()),
+	}
+	tag2 := kv.Pair{
+		Key:   []byte("tx.validator"),
+		Value: st.ValidatorAddress.Bytes(),
+	}
+	tag3 := kv.Pair{
+		Key:   []byte("tx.delegator"),
+		Value: st.StakeAddress.Bytes(),
+	}
+	tag4 := kv.Pair{
+		Key:   []byte("tx.amount"),
+		Value: st.Stake.Value.BigInt().Bytes(),
+	}
+
+	tags = append(tags, tag, tag2, tag3, tag4)
```

Comment:
Put all fields to tags?

Reply:
NodeName, ValidatorECDSAPubKey and ValidatorPubKey - redundant and could be taken from leveldb.